Create Zone -- APIs

Jay07GIT opened the pull request:
Fixes #1455.

Changes in this pull request:
User can generate the New zone in vinyldns.

Create- Post request :- `http://localhost:9000/zones/generate`
Update- Put request :- `http://localhost:9000/zones/generate`
Delete- Delete request :- `http://localhost:9000/zones/generate/<"zone_id">`
List- Get Request :- `http://localhost:9000/zones/generate/info?maxItems=100&nameFilter=<"name">&ignoreAccess=false `
List by name- Get Request :- `http://localhost:9000/zones/generate/name/"zone_name"`
List by ID- Get Request :- `http://localhost:9000/zones/generate/id/"zone_name"`

Create Zone request - 
for powerdns
```
{
    "groupId": "groupid",
    "provider": "powerdns",
    "zoneName": "mttest1.example.org.", 
    "email": "test@test.com",
    "providerParams":{
        "kind": "Native", 
        "nameservers": ["172.17.42.1.", "ns1.mttest1.example.org."]}
        "masters": ["172.17.42.1", "172.17.42.2"]
    }
}
```
for bind -  "ns_ipaddress": ["172.17.42.1"], use the same to test the connection
```
{
    "groupId": "groupid",
    "provider": "bind",
    "zoneName": "test.com.",
    "providerParams":{
        "nameservers": ["172.17.42.1"],
        "admin_email": "admin@test.com",
        "ttl": 3600,
        "refresh":  6048000,
        "retry": 86400,
        "expire":  24192000,
        "negative_cache_ttl":  6048000
    }
}
```
To test the bind server zone creation in vinyldns ,

Comment the image in the file `quickstart/docker-compose.xml`
#image : "vinyldns/build:base-test-integration-${VINYLDNS_IMAGE_VERSION}"

run 
```
./quickstart/quickstart-vinyldns.sh --clean
./quickstart/quickstart-vinyldns.sh --deps-only 
```

Update zone request:- 
```
{
    "groupId": "b82c2a36-a973-4abd-b47d-0711b105c0d5",
    "email":"test@test.com",
    "provider": "powerdns",
     "zoneName": "mttest13.example.org.",
     "providerParams":{
         "kind": "Master"
      }
}
```

Dependency PR: (Merge before this PR)
 for the protobuf update.:- https://github.com/vinyldns/vinyldns/pull/1207
 for the logs.:- https://github.com/vinyldns/vinyldns/pull/1446

To do
Update zone for the test bind server.

nspadaccino committed:
generate zone route

Jay07GIT committed:
created api for zone create with validations

Jay07GIT committed:
Added create zone api in config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Added bind validations

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
json request formatted

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added api key config for zone creation

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
formatted powerdns create zone response

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added create zone api in bind test docker image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added db validations

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
insert values in generate zone db

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added duplicate validation on generate zones

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added pdns config for connect zone

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated generate zones config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
removed unwanted files

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added created timestamp

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added created timestamp test

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add pdns in quickstart

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update pdns config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update pdns config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added list created zones api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added email field in create zone api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added access level field in create zone api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
use json4s to build zone generate requests

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve failed tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve failed tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated generate_zones_bind_api.py

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated manage_vinyldns_zones_bind_api.py with delete zones api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated conf with delete bind uri

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

nspadaccino committed:
zone create provider templates in config

Jay07GIT committed:
updated the get generated zone

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added list generated zone tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

nspadaccino committed:
updated zone generation input and protobuf, added required fields to config

nspadaccino committed:
updates for provider-agnostic zone generation

nspadaccino committed:
update comments

nspadaccino committed:
prune unused optional params from template, enhanced validation and error handling

nspadaccino committed:
Use json schema for validating provider api requests

nspadaccino committed:
convert response message from json string to json

nspadaccino committed:
updated template engine

nspadaccino committed:
add delete and update endpoints, handler updates, added endpoint string substitution

nspadaccino committed:
update zone generation handler, repository methods

nspadaccino committed:
added delete and update generated zone

Jay07GIT committed:
removed the unused commented code

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
upgraded the slf4j-api and logback-classic version to latest

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
updating from logback to log4j2 due to version mismatch in slf4j-api

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
undo illegal header config

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated create zone params

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve existing tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve existing tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
add tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added it tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added it tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added get zones by id

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'master' into create-zone-api

nspadaccino committed:
Merge branch 'master' into logs_dependency_update

nspadaccino committed:
Merge remote-tracking branch 'jay/logs_dependency_update' into create-zone-api

nspadaccino committed:
updated dependencies

Jay07GIT committed:
added functional tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
added functional tests for update,get,delete,list zones

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update func docker network

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'version-0.22.0-beta.1' into create-zone-api

Jay07GIT committed:
trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Merge branch 'version-0.22.0-beta.1' into create-zone-api

Jay07GIT committed:
resolve tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolve tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'create-zone-api' of github.com:nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
Merge branch 'create-zone-api' of https://github.com/nspadaccino/vinyldns into create-zone-api

Jay07GIT committed:
resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

resolved play tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

changed the pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

changed the pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

update for pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

update for pdns image

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added deleted zone for bind

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
Merge branch 'main' into create-zone-api

nspadaccino committed:
fix typo

nspadaccino committed:
fix typo

JoshSEdwards commented on modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala:
```diff
@@ -135,6 +159,394 @@ class ZoneService(
 
+  def getGenerateZoneByName(zoneName: String, auth: AuthPrincipal): Result[GenerateZone] =
+    for {
+      generateZone <- getGenerateZoneByNameOrFail(ensureTrailingDot(zoneName))
```
I think the generated zone read paths are missing the access check this PR introduced. `canSeeGenerateZone` exists, but `getGenerateZoneByName` and `getGeneratedZoneById` return the generated zone directly. Can we require `canSeeGenerateZone(auth, generatedZone)` before returning those reads?

I also don't love that some methods and vars are missing the D at the end of generated.

Jay07GIT replied:
Done the changes

nspadaccino committed:
Merge branch 'main' into create-zone-api

JoshSEdwards commented on modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala:
```diff
@@ -135,6 +159,394 @@ class ZoneService(
+        generatedZones = listZonesResult.generatedZones
+        groupIds = generatedZones.map(_.groupId).toSet
+        groups <- groupRepository.getGroups(groupIds)
+        generateZoneSummaryInfos = generateZoneSummaryInfoMapping(generatedZones, authPrincipal, groups)
```
This looks like it fetches all matching generated zones first and only computes `accessLevel` after the data has already been loaded into the response model. The repository signature accepts `authPrincipal` and `ignoreAccess` but the MySQL implementation doesn't use either one to filter results. We should probably enforce access before building the response.

Jay07GIT replied:
Added the required code.

JoshSEdwards commented on modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala:
```diff
@@ -135,6 +159,394 @@ class ZoneService(
+      )
+      zoneToUpdate = existingGeneratedZone.copy(
+        email = request.email,
+        groupId = request.groupId,
```
This allows the generated-zone update flow to reassign ownership without checking access to the target group. The normal zone update path checks both the current admin group and the requested new admin group, so this should probably follow the same rules.

Jay07GIT replied:
Resolved

JoshSEdwards commented on modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneValidations.scala:
```diff
@@ -54,6 +54,11 @@ class ZoneValidations(syncDelayMillis: Int) {
 
+  def isValidGenerateZoneConn(responseCode : Int, responseMsg : String): Either[Throwable, Unit] =
+    ensuring(InvalidRequest(responseMsg)) {
+      responseCode <= 400
```
This should probably be an explicit 2xx check, or at least `< 400` so it doesn't include a 400 response as a success.

Jay07GIT replied:
Resolved

JoshSEdwards commented on modules/core/src/main/scala/vinyldns/core/domain/zone/Zone.scala:
```diff
@@ -123,8 +140,78 @@ object Zone {
     )
   }
 }
+final case class GenerateZone(
```
The core `Zone.scala` file now contains a lot of generated zone and provider-integration code. `GenerateZone`, generated-zone request/response transport models, json4s payloads, provider endpoint templates, API keys, and provider config parsing.

Feels like feature-specific integration is leaking into the canonical zone domain model here. Does it make more sense to move the generated zone models/config into a dedicated module or file so the core zone domain stays focused?

nspadaccino replied:
Yeah it definitely makes sense to refactor this into its own module

Jay07GIT replied:
Since we do not have a separate Connect Zone file, we commonly use the Zone file for that purpose. Therefore, it would be consistent to place generateZone within zones as well, since it is part of the Zone configuration.

JoshSEdwards commented on modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlGenerateZoneRepository.scala:
```diff
@@ -0,0 +1,257 @@
+      }
+    }
+
+  def listGenerateZones(
```
`listGenerateZones` and `listGeneratedZonesByAdminGroupIds` look like they're auth-aware from the arguments they take, but the MySQL implementations don't use `authPrincipal` or `ignoreAccess` anywhere. We should either use them, or if we're intentionally keeping auth out of this remove them from the args list.

Jay07GIT replied:
Resolved. Removed the authPrincipal which is not in use.

JoshSEdwards commented on modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlGenerateZoneRepository.scala:
```diff
@@ -0,0 +1,257 @@
+
+          val filters = if (zoneNameFilter.isDefined && (zoneNameFilter.get.takeRight(1) == "." || zoneNameFilter.get.contains("*"))) {
+            List(
+              zoneNameFilter.map(flt => s"gz.name LIKE '${ensureTrailingDot(flt.replace('*', '%'))}'"),
```
This query is being built by interpolating request-derived values directly into SQL. `zoneNameFilter` and `startFrom` are turned into SQL fragments here and then appended into the final query with `filters.mkString(" AND ")`.

Can we use bind parameters for these values instead? Even if similar patterns exist elsewhere, this is new code and should avoid carrying that risky design pattern forward.

Jay07GIT replied:
Resolved.

JoshSEdwards reviewed (changes requested):
A few comments, mostly around auth, but also I'm a little concerned with mixing all the new generated-zone code in the same files as existing zone code, especially Zone.scala. The feature is spread through routing, service, core domain, config loading, and persistence, all without a clear boundary. I realize that's asking for a big refactor, so I'm open to hearing why it should remain as-is.

Sokitha-21 commented on modules/api/src/main/scala/vinyldns/api/domain/zone/ZoneService.scala:
```diff
@@ -417,6 +843,35 @@ class ZoneService(
 
+  private def generateZoneDoesNotExist(zoneName: String): Either[Throwable, Unit] = {
+    val existingZoneOpt: Option[GenerateZone] =
+      generateZoneRepository.getGenerateZoneByName(zoneName).unsafeRunSync()
```
unsafeRunSync() is called inside a method that is already executing within a for. Calling unsafeRunSync() on a thread that is already managed by the runtime can cause a deadlock.

Jay07GIT replied:
Resolved.

Sokitha-21 commented on modules/api/src/main/scala/vinyldns/api/route/ZoneRouting.scala:
```diff
@@ -143,6 +142,91 @@ class ZoneRoute(
+      }
+    } ~
+    path("zones" /"generate"/ "name" / Segment) { zoneName =>
+      get & monitor("Endpoint.getGenerateZoneByName")
```
The braces {} are missing. This means the route is not restricted to GET requests and the monitor is never applied.

Jay07GIT replied:
Resolved.

Sokitha-21 commented on modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlGenerateZoneRepository.scala:
```diff
@@ -0,0 +1,257 @@
+
+          if(startFrom.isDefined){
+            sb.append(" AND ")
+            sb.append(s"gz.name > '${startFrom.get}'")
```
startFrom is interpolated directly as a string literal and adminGroupIds values are wrapped in single quotes and joined into the query. Can we use bind parameters as suggested before instead of manual string construction.

Jay07GIT replied:
Resolved.

Sokitha-21 commented on modules/mysql/src/main/scala/vinyldns/mysql/repository/MySqlGenerateZoneRepository.scala:
```diff
@@ -0,0 +1,257 @@
+                  'name -> generateZone.zoneName,
+                  'provider -> generateZone.provider,
+                  'adminGroupId -> generateZone.groupId,
+                  'response -> toPB(generateZone.response.get).toByteArray,
```
response is Option[ZoneGenerationResponse]. Calling .get will throw NoSuchElementException at runtime whenever a GenerateZone is saved without a response.

Jay07GIT replied:
Resolved.

Sokitha-21 commented on modules/api/src/main/scala/vinyldns/api/route/ZoneRouting.scala:
```diff
@@ -143,6 +142,91 @@ class ZoneRoute(
+        }
+      }
+    } ~
+    path("zones" / "generate" / Segment) { id =>
```
Segment acts as a wildcard and can match routes like info, allowedDNSProviders and nameservers causing requests to hit the wrong route and return 405 Method Not Allowed.
Can we consider moving the specific routes above the wildcard Segment routes so they are matched first.

Jay07GIT replied:
Resolved.

Jay07GIT commented on project/plugins.sbt:
```diff
@@ -20,7 +20,7 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-ghpages" % "0.6.3")
 addSbtPlugin("de.heikoseeberger" % "sbt-header" % "5.5.0")
 
-addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.7.4")
+addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.8.18")
```
@arpit4ever These deps update are part of the news deps added. So its require to run this PR success.

Jay07GIT committed:
resolved the review comments

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
updated the sql syntax

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
resolved func tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
trigger tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added auth tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
added auth tests

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

Jay07GIT committed:
update

Signed-off-by: Jay07GIT <jeyraj931@gmail.com>

nspadaccino committed:
security: enforce access control on generated-zone list endpoints

  listGenerateZones and listGeneratedZonesByAdminGroupIds returned every
  generated zone in the system regardless of the caller's group membership:
  the repository methods took no AuthPrincipal and never applied an access
  filter, so the `ignoreAccess` flag was inert and any authenticated user
  could enumerate all generated zones.

  Generated zones have no per-zone ACL table (unlike regular zones), so
  access is purely admin-group membership. Add an accessFilter helper that
  mirrors MySqlZoneRepository.withAccessors semantics: super/support users,
  or callers explicitly passing ignoreAccess, see all zones; everyone else
  is restricted to zones owned by a group they belong to; a user in no
  groups sees nothing. Thread AuthPrincipal through the repository algebra
  and ZoneService into both list queries.

  Add integration regression tests covering: non-members cannot see another
  group's zones (direct and admin-group-search paths), members and super
  users can, and ignoreAccess restores the unfiltered view.

nspadaccino committed:
security: encrypt provider API key instead of holding it in plaintext

  DnsProviderConfig.apiKey was a plain String loaded from config and passed
  around in memory, diverging from how every other connection secret in
  VinylDNS is handled (Encrypted via CryptoAlgebra).

  Change apiKey to Encrypted, encrypt it at config-load time, and decrypt
  it just-in-time at the three ZoneService call sites only when setting the
  X-API-Key header. The in-memory representation is now ciphertext, so the
  key cannot leak through accidental logging or toString.

nspadaccino committed:
security: drop server-owned fields from generate-zone request model

  ZoneGenerationInput (the generate-zone request body) carried id, status,
  and response with defaults, and the JSON deserializer read them straight
  from the request. A client could therefore set a zone's id/status or
  inject a fabricated provider response, which then flowed into the
  persisted GenerateZone.

  Remove id, status, and response from ZoneGenerationInput and stop reading
  them in ZoneGenerationInputSerializer. The server now owns these fields:
  GenerateZone assigns a fresh id and Active status, and ZoneService sets
  the response from the actual provider call. Add a routing regression test
  that posts id/status in the body and asserts they are ignored.

nspadaccino committed:
security: fail closed when no provider schema is configured

  schemaValidationResult returned success when a provider had no JSON schema
  for an operation, so a provider misconfigured without a create-zone or
  update-zone schema would forward unvalidated providerParams straight to
  the external provider.

  Reject in that case instead. A missing schema is an operator
  misconfiguration rather than bad client input, so surface it as a
  server-side failure (500). Add a service test asserting a provider with
  no schema is rejected before any provider call.

nspadaccino committed:
security: bound external DNS-provider calls with timeouts

  createConnection opened an HttpURLConnection with no connect or read
  timeout, so a slow or hung provider would block the request thread
  indefinitely. Set a 10s connect and 30s read timeout. A configurable
  client abstraction is left for the follow-up refactor.

nspadaccino committed:
test: update generated-zone list functional tests for access scoping

  Phase 0 added access control to the generated-zone list endpoints, so a
  caller now sees only zones owned by groups they belong to. Four functional
  tests in list_generated_zones_test.py asserted counts/pagination from the
  old global-visibility behavior (e.g. has_length(7)/(8)) and failed in CI.

  Update them to the access-scoped 5-zone set owned by list_generate_zones_group,
  matching the already-passing second_page test.

nspadaccino committed:
review: address PR feedback on generated-zone hardening

  Three changes from Josh's review:

  - ignoreAccess: remove the parameter entirely from the generated-zone list
    endpoints instead of leaving it as a bypass. Access is now role-based —
    super/support admins see all generated zones, every other caller is always
    scoped to the groups they belong to. Drops the query param, the repository/
    service params, and the ignoreAccess field from ListGeneratedZonesResults
    and ListGeneratedZonesResponse.

  - providerParams: reject a request whose providerParams is present but not a
    JSON object (400 "providerParams must be a JSON object") instead of quietly
    collapsing the invalid payload into an empty map.

  - access filter: cap the member-group accessor list at MAX_ACCESSORS (30) with
    a warning, mirroring MySqlZoneRepository.buildZoneSearchAccessorList, to
    avoid building an unbounded parameterized query.

  Tests updated accordingly: ZoneServiceSpec/ZoneRoutingSpec/integration specs no
  longer thread ignoreAccess; added a routing test for malformed providerParams;
  functional list tests now assert access-scoped results with admins seeing all.

nspadaccino committed:
Merge pull request #1 from nspadaccino/czapi-phase0-security

Phase 0: security hardening for the generated-zone API
